Add support for incoming webhook

chmouel opened the pull request:
let the user specify a shared secret and be able to target a
PipelineRun from a repo crd match.

We only let the user target what is explicitely targetted in the Repo
CRD, just in case you know for security being explicite is better to be
implicit.

There is room for improvement in the future. We currently do
webhook-url type which is not used anywhere, but we could have a slack
or other methods to be able to answer to slack if needed.

TODO: need to add an E2E test

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Add support for incoming webhook

let the user specify a shared secret and be able to target a
PipelineRun from a repo crd match.

We only let the user target what is explicitely targetted in the Repo
CRD, just in case you know for security being explicite is better to be
implicit.

There is room for improvement in the future. We currently do
webhook-url type which is not used anywhere, but we could have a slack
or other methods to be able to answer to slack if needed.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

sm43 commented on pkg/params/info/events.go:
```diff
@@ -20,6 +21,9 @@ type Event struct {
 	TriggerTarget string
 
+	// Target PipelineRun, the target PipelineRun user request. Used in incoming webhook
+	TargetPipelineRun string
```
do we want to use https://github.com/openshift-pipelines/pipelines-as-code/blob/99afbaa5342f8559227580fa97bef44a8522b1f8/pkg/params/info/events.go#L59 for this?
or need a new one

chmouel commented:
Ah yeah that’s maybe better or we can have state.TargetTestPipelineRun to use this, since it make more sense I think to be in the top-level? Wdyt?

> On 30 May 2022, at 15:40, Shivam Mukhade ***@***.***> wrote:
> 
> 
> @sm43 commented on this pull request.
> 
> In pkg/params/info/events.go <https://github.com/openshift-pipelines/pipelines-as-code/pull/709#discussion_r884843686>:
> 
> > @@ -20,6 +21,9 @@ type Event struct {
>  	// a push or a pull_request
>  	TriggerTarget string
>  
> +	// Target PipelineRun, the target PipelineRun user request. Used in incoming webhook
> +	TargetPipelineRun string
> do we want to use https://github.com/openshift-pipelines/pipelines-as-code/blob/99afbaa5342f8559227580fa97bef44a8522b1f8/pkg/params/info/events.go#L59 <https://github.com/openshift-pipelines/pipelines-as-code/blob/99afbaa5342f8559227580fa97bef44a8522b1f8/pkg/params/info/events.go#L59> for this?
> or need a new one
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/openshift-pipelines/pipelines-as-code/pull/709#pullrequestreview-989371059>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAYFJB5YQIQZBTIMWRWBGTVMTANPANCNFSM5XKRZLFA>.
> You are receiving this because you authored the thread.
>

sm43 commented:
> yeah `TargetPipelineRun` is fine. but do we want to breakdown Event{} 🤔 

I want to but I think that one is fine, the provider specific variable is the annoying ones (which was just plain lazyness for me )...

sm43 commented on docs/content/docs/guide/incoming_webhook.md:
````diff
@@ -0,0 +1,75 @@
+`https://github.com/owner/repo`. As an example here is the full curl snippet:
+
+```shell
+curl -X POST https://control.pac.url/incoming?secret=very-secure-secret&repository=repo,branch=main
````
we are passing actual secret here 🤔 
i was thinking we would pass secret name and pac will use the token from there

chmouel replied:
That's the webhook secret, the same way we have a webhook secret on other methods... it's actually a pattern for webhooks based URL.

chmouel replied:
fyi: this engineer explain it well the different method : 

https://code.likeagirl.io/best-practice-to-secure-your-webhooks-618684813723

I do want to do body hmac validation like done on github payload webhook, but since we don't have body payload on incoming webhook we can't hmac much

sm43 commented:
if we could get https://github.com/openshift-pipelines/pipelines-as-code/pull/706 in before this

sm43 commented:
> TODO: need to add an E2E test

are you gonna address this in another pr?
we can merge this

chmouel commented:
yeah i'll do the e2e part of the pr, just need to be properly online for this  (still traveling)

chmouel commented:
stil want to add e2e test but want to target gitlab for the e2e this time and due have to  add support for gitlab :)

chmouel committed:
allow redefining the prot of the watcher/controller

makes it easy to debug from the same debugger

chmouel committed:
add support for gitlab webhook-incoming

chmouel committed:
add gitlab e2e on incoming-webhook

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Add repo-url annotation to generated PipelineRun

Add the repo url to get the real repo URL where it's coming from.  We could do
this as well with event.Organization and event.Repository but since those are
labels they get transformed to please lord k8s and most of the time that's fine
but with gitlab you can have / in "Organization" (which is actually called
Project there) and since they / gets transformed to - by pkg/formatting/k8labels
we would not know the proper name.

We actually do need the proper name since we would not know where it's coming
from to update status with gitlab.

so now the reconciler/watcher will use that value to fill up the events.org and
events.repo. Added bunch of test to handle all use case i can think of.

Acked-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented on pkg/cli/webhook/github_test.go:
```diff
@@ -46,7 +46,7 @@ func TestAskGHWebhookConfig(t *testing.T) {
 				as.StubOne("token")
 			},
-			repoURL:       "https:/github.com/pac/demo",
+			repoURL:       "https://github.com/pac/demo",
```
was this test a bad url on purpose? because we know validate the URL properly in formatting.GetRepoOwnerURL and it was failing, i wasn't sure of the test.

sm43 replied:
wasn't on purpose 🤔 it didn't failed but

chmouel commented on pkg/apis/pipelinesascode/v1alpha1/types.go:
```diff
@@ -61,16 +61,24 @@ type RepositoryRunStatus struct {
 }
 
-type GitProviderSecret struct {
+type Secret struct {
```
renamed it because it's not tight to gitprovider anymore but for other secrets...
